include/swift/AST/DiagnosticList.h

```diff
@@ -32,6 +33,11 @@ enum class DiagID : uint32_t {
 static_assert(static_cast<uint32_t>(swift::DiagID::invalid_diagnostic) == 0,
               "0 is not the invalid diagnostic ID");
 
+constexpr const std::underlying_type_t<DiagID> NumDiagIDs = 0
+#define DIAG(KIND, ID, Group, Options, Text, Signature) +1
+#include "swift/AST/DiagnosticsAll.def"
+    ;
```

Comment:
Can't we add `NumDiags` member to `DiagID` instead?

Reply:
I think we can, but we would have to static_cast it to an integer every time because of the `enum class`.

Reply:
Maybe that's better than a gigantic string + 1s in the header?

Reply:
Ah, good point. Having to cast is not a great user experience, though. Can I win you over with an `extern`?

Reply:
Sure, let's at least do that.

Reply:
Not gonna work, we need a `constexpr`.